### PR TITLE
[StructuralMechanicsApplication][TO DISCUSS+NOT MERGE] Adding RuleOfMixturesLaw (composite parallel law)

### PR DESCRIPTION
**Pull request by loumalouomega**
This uses everything added on #2919. These two PR are related and they work together, one change on #2919 must be applied in this too

This PR adds a composite parallel law (classic rule of mixtures). This work is in close collaboration with @frastellini, he is more familiarized with the technical details

@frastellini, here we can discuss how to consider geometries with more than one GP

### CODE REVIEW

**Review by frastellini — DISMISSED** [review]
Perfect! Thank you

**Comment by loumalouomega** [comment]
Let's wait for the @KratosMultiphysics/technical-committee  to approve the #2919

**Review by pooyan-dadvand — DISMISSED** [review]
As it is approved but title says for discussion, I block it to avoid merge by mistake.

**Comment by loumalouomega** [comment]
Can we push this a bit?, I am solving too many conflicts

**Comment by loumalouomega** [comment]
Conflicts solved

**Comment by loumalouomega** [comment]
@frastellini can you please contact with @AlejandroCornejo, he is interested in continue the SP developments.

**Comment by frastellini** [comment]
Yes. We have already commented the current state of developments and I offered Alejandro all my help for the joint development of Serial-Parallel mixing theory in Kratos (after 13 years of the doctoral thesis).
The complete SP composite model is very expensive in terms of computational cost. We can implement the complete model, but also cheaper models like the Vanishing-Fibre-Diameter (where the fibre only contributes in fibre direction, and neglected in transversal ones)

File: applications/StructuralMechanicsApplication/custom_constitutive/rule_of_mixtures_law.cpp [hunk 1]
```diff
@@ -0,0 +1,1497 @@
...
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+bool RuleOfMixturesLaw::Has(const Variable<bool>& rThisVariable)
+{
+    // At least one layer should have the value
+    bool has = false;
+    KRATOS_DEBUG_ERROR_IF(mConstitutiveLaws.size() == 0) << "RuleOfMixturesLaw: No constitutive laws defined" << std::endl;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/rule_of_mixtures_law.cpp]: maybe makes more sense to do once in `Check`?

**Review by philbucher — COMMENTED** [review]
can you please move this law to the advanced-laws folder?

Will you create a test for it?

**Comment by loumalouomega** [comment]
Done

**Review by philbucher — CHANGES_REQUESTED** [review]
please move the test to the advanced claws tests folder

**Comment by philbucher** [comment]
> please move the test to the advanced claws tests folder

same for the law itself, still in the wrong folder

Otherwise ok for me

**Review by philbucher — DISMISSED** [review]
ok for me
I guess the CLaw guys had enough time to review it ;)
ping @pooyan-dadvand

**Comment by loumalouomega** [comment]
> 
> 
> ok for me
> I guess the CLaw guys had enough time to review it ;)
> ping @pooyan-dadvand

Done, I moved the files

**Comment by loumalouomega** [comment]
@pooyan-dadvand can you release the block?, @AlejandroCornejo can you review this?

**Comment by loumalouomega** [comment]
Pleople from @KratosMultiphysics/structural-mechanics @AlejandroCornejo can anyone review this?

**Review by AlejandroCornejo — APPROVED** [review]
looks okkey to me, there's any test that I can check in order to update the PR of the serial-parallel RoM?

**Pull request merged**